lib/ts/ArgParser.cc

```diff
@@ -0,0 +1,506 @@
+{
+  // handle the args
+  if (arg_num == INFINITE_ARG_NUM) {
+    // infinite arguments
+    for (unsigned long j = index + 1; j < args.size(); j++) {
+      data.arg_data.push_back(args[j]);
+    }
+    args.erase(args.begin() + index, args.end());
+    ret.append(name, data);
+    return;
+  }
+  // finite number of argument handling
+  for (int j = 0; j < arg_num; j++) {
+    if (args.size() < index + j + 2 || args[index + j + 1].empty() || args[index + j + 1][0] == '-') {
+      std::cout << "Error: " << arg_num << " argument(s) expected by " << name << std::endl << std::endl;
+      base.help_message();
+    }
+    data.arg_data.push_back(args[index + j + 1]);
+  }
+  // erase the used arguments and append the data to the return structure
+  args.erase(args.begin() + index, args.begin() + index + arg_num + 1);
+  ret.append(name, data);
+  index -= 1;
+  return;
+}
+
+// append the args of option to parsed data
+static void
+append_option_data(ArgParser &base, ParsedArgs &ret, StringArray &args,
+                   std::unordered_map<std::string, ArgParser::Option> const &option_list,
+                   std::unordered_map<std::string, std::string> const &option_map, int index)
+{
+  for (unsigned long i = index; i < args.size(); i++) {
+    // find matches of the arg
+    if (args[i][0] == '-' && args[i][1] == '-' && args[i].find('=') != std::string::npos) {
+      // deal with --args=
+      std::string option_name = args[i].substr(0, args[i].find_first_of('='));
+      if (option_list.find(option_name) != option_list.end() && option_list.at(option_name)._opt_arg_num == 1) {
+        ParserData option_data;
+        ArgParser::Option cur_option = option_list.at(option_name);
+        option_data.arg_data.push_back(args[i].substr(args[i].find_last_of('=') + 1));
+        args.erase(args.begin() + i);
+        ret.append(cur_option._opt_name, option_data);
+        i -= 1;
+      }
+    } else if (option_list.find(args[i]) != option_list.end() || option_map.find(args[i]) != option_map.end()) {
+      // arg name match or key match
+      // output version or help message
+      if (args[i] == "--version" || args[i] == "-V") {
+        base.version_message();
+      }
+      if (args[i] == "--help" || args[i] == "-h") {
+        base.help_message();
+      }
+      ParserData option_data;
+      ArgParser::Option cur_option;
+      if (option_list.find(args[i]) != option_list.end()) {
+        cur_option = option_list.at(args[i]);
+      } else {
+        cur_option = option_list.at(option_map.at(args[i]));
+      }
+      handle_args(base, ret, args, option_data, cur_option._opt_name, cur_option._opt_arg_num, i);
+    }
+  }
+}
+
+// Main recursive logic of Parsing
+void
+ArgParser::Command::parse(ArgParser &base, ParsedArgs &ret, StringArray &args)
+{
+  // check for conflict commands
+  int conflict_flag = 0;
+  for (unsigned long i = 0; i < args.size(); i++) {
```

Comment:
In cases where the index isn't externally visible, use a range for loop. Only use indexed loops when there's a specific need to do so.